isc_joy/src/joystick.cpp

```diff
@@ -0,0 +1,136 @@
+			n.getParamCached("mappings/" + *control + "/type", tmp); // cache the params and we don't care about the actual value yet
+			n.getParamCached("mappings/" + *control + "/index", tmp); 
+		} else { // if either one is missing, this control is malformed and will not be reported
+			n.setParam("mappings/" + *control + "/type", -1);
+			n.setParam("mappings/" + *control + "/index", -1);
+
+			ROS_WARN("Malformed or missing mapping for control %s. Control value will be set to 0", control->c_str()); 			
+
+			n.getParamCached("mappings/" + *control + "/type", tmp); // even if a control is missing or malformed, we must cache it to avoid 
+			n.getParamCached("mappings/" + *control + "/index", tmp); // the param server lookup
+		}
+	}
+}
+
+float get_mapping(std::string control, const sensor_msgs::Joy::ConstPtr &joy) {
+	ros::NodeHandle n("~");
+	control_type type;
+	int index;
+	
+	n.getParamCached("mappings/" + control + "/type", type); // since it's cached the lookup is actually very quick
+	n.getParamCached("mappings/" + control + "/index", index);
+
+	ROS_INFO("%s: type=%d, index=%d", control.c_str(), type, index);
+
+	switch(type) {
+		case CONTROL_TYPE_BUTTON:
+			if(index < joy->buttons.size() && index >= 0) { return joy->buttons[index]; } // do some quick bounds checking too, just in case index is wrong
+			else {
+				ROS_WARN("%s: index %d is out of bounds", control.c_str(), index);
+				return 0.0;
+			}
+			break;
+
+		case CONTROL_TYPE_AXIS:
+			if(index < joy->axes.size() && index >= 0) { return joy->axes[index]; }
+			else {
+				ROS_WARN("%s: index %d is out of bounds", control.c_str(), index);
+				return 0.0;
+			}
+			break;
+
+		default: // if this control was improperly defined or not defined at all, just report 0 for the control
+			ROS_WARN("%s is badly defined.", control.c_str());			
+			return 0.0;
+	}
+}
+
+void joystickCallback(const sensor_msgs::Joy::ConstPtr& joy){	
+	/* This fires every time a button is pressed/released
+	and when an axis changes (even if it doesn't leave the
+	deadzone). */
+	
+	isc_joy::xinput msg;
+	msg.A = get_mapping("A", joy);
+	msg.B = get_mapping("B", joy);
+	msg.X = get_mapping("X", joy);
+	msg.Y = get_mapping("Y", joy);
+	msg.LB = get_mapping("LB", joy);
+	msg.RB = get_mapping("RB", joy);
+	msg.Back = get_mapping("BACK", joy);
+	msg.Start = get_mapping("START", joy);
+	msg.Guide = get_mapping("GUIDE", joy);
+	msg.LS = get_mapping("LS", joy);
+	msg.RS = get_mapping("RS", joy);
+
+	msg.LeftStick_LR = get_mapping("L_LR", joy);
+	msg.LeftStick_UD = get_mapping("L_UD", joy);
+	msg.RightStick_LR = get_mapping("R_LR", joy);
+	msg.RightStick_UD = get_mapping("R_UD", joy);
+	msg.LT = get_mapping("LT", joy);
+	msg.RT = get_mapping("RT", joy);
+	msg.DPad_LR = get_mapping("DPAD_LR", joy);
+	msg.DPad_UD = get_mapping("DPAD_UD", joy);
+
+	joystickPub.publish(msg);
+}
+
+int main(int argc, char **argv){
+	ros::init(argc, argv, "joystick_xbox360");
```

Comment:
The node name needs to be updated.

Reply:
Unless we don't really care, I'm leaving the node type as is so to not break older launch files. I still have to add in a default rosparam load, though.

Reply:
Hmm I see your comment in the CMake now. I think we would just mark the current version as a [release on GitHub](https://github.com/iscumd/SharedROSNodes/releases) so that older code can use that, but we're changing the name for what wold be version 3.0.0 (this isn't the first time we've done this). If it's still a concern, we could create an extra node that's a shim around the new node that keeps the old naming.